Update OrdinationResults class

ElDeveloper opened the pull request:
**Do not merge**

Fixes #521

to-do before merge:

- [x] update PCoA method and tests (@jairideout)
- [x] update RDA method and tests (@mortonjt)
- [x] update CA method and tests (@ElDeveloper)
- [x] update CCA method and tests (@josenavas / @mortonjt)
- [x] fix broken `OrdinationResults.__str__` method and update tests to catch this failure
- [x] fix coverage issues (definitely `skbio.util.assert_ordination_results_equal`, possibly others). Rewrite unit tests to cover new keyword arguments (@ebolyen)
- [x] test methods against vegan package test data. Important: make sure the constrained methods (CCA, RDA) either 1) report *only* constrained axes **OR** 2) if they report both constrained and unconstrained axes, the axes are labeled correctly.
- [x] fix RDA eigenvalues bug (@Jorge-C)
- [x] get all tests to pass
- [x] investigate using `scipy.linalg` wherever `numpy.linalg` is used (see [here](http://docs.scipy.org/doc/scipy/reference/tutorial/linalg.html#scipy-linalg-vs-numpy-linalg) for reasons why we want to do this)
- [ ] detailed review from anyone available
- [ ] ask @Jorge-C to review specific sections of the docs (ideally all changes if possible). We'll add comments to specific parts of the docs that need review

Jorge-C commented:
Git question: Is it possible to make git understand code movement, so that it can still track its history?

ElDeveloper commented:
@Jorge-C from what I can tell, no there's no way to do that.

On (May-13-15|12:30), Jorge Cañardo Alastuey wrote:
> Git question: Is it possible to make git understand code movement, so that it can still track its history?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/biocore/scikit-bio/pull/925#issuecomment-101785701

ElDeveloper commented on skbio/util/_testing.py:
```diff
@@ -94,3 +97,81 @@ def get_data_path(fn, subfolder='data'):
+    compare ``OrdinationResults`` objects.
+
+    For numeric attributes (e.g., eigvals, site, etc.),
+    ``numpy.testing.assert_almost_equal`` is used. Otherwise,
```
I believe this is out of date.

jairideout replied:
fixed, thx

jairideout commented:
@josenavas @ElDeveloper are either of you available to finish off the remaining to-do items?

josenavas commented:
I will be able to do something after ASM...

ElDeveloper commented:
I think I should be able to have a look later next week or early the
week after.

On (May-20-15| 9:37), Jai Ram Rideout wrote:
> @josenavas @ElDeveloper are either of you available to finish off the remaining to-do items?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/biocore/scikit-bio/pull/925#issuecomment-103952365

jairideout commented:
Awesome, thanks guys! I likely won't have time to work on this either until after ASM.

mortonjt commented:
If you guys need an extra hand, I'll be happy to clean up the testing suite
and add vegan test cases
On May 22, 2015 7:22 PM, "Jai Ram Rideout" <notifications@github.com> wrote:

> Awesome, thanks guys! I likely won't have time to work on this either
> until after ASM.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/biocore/scikit-bio/pull/925#issuecomment-104810751>.
>

ElDeveloper commented:
That'd be great!

Yoshiki Vázquez-Baeza

> On May 22, 2015, at 10:54 PM, mortonjt <notifications@github.com> wrote:
> 
> If you guys need an extra hand, I'll be happy to clean up the testing suite
> and add vegan test cases
> On May 22, 2015 7:22 PM, "Jai Ram Rideout" <notifications@github.com> wrote:
> 
> > Awesome, thanks guys! I likely won't have time to work on this either
> > until after ASM.
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/biocore/scikit-bio/pull/925#issuecomment-104810751>.
> >
> —
> Reply to this email directly or view it on GitHub.
>

jairideout commented:
Thanks @mortonjt!

jairideout commented:
@mortonjt any updates on your end? We're shooting to have this PR included in the upcoming beta release (0.4.0) on June 30.

mortonjt commented:
Not too good.  The output from RDA is not consistent with vegan.

I'm meeting up with @JorgeC Wednesday.  Will give update afterwards

@mortonjt <https://github.com/mortonjt> any updates on your end? We're
shooting to have this PR included in the upcoming beta release (0.4.0) on
June 30.

—
Reply to this email directly or view it on GitHub
<https://github.com/biocore/scikit-bio/pull/925#issuecomment-112180592>.

Jorge-C commented:
That's odd, because RDA is explicitly tested against results computed with vegan 2.0-8, as mentioned in its test class, though maybe things have changed in more recent versions (?).

mortonjt commented:
Here's a test case where I see differing output in the original code (so before this PR).  The test files are from the vegan manual is attached in an email to the mailing list.  Also, this is all using vegan 2.0-8

```
from skbio.stats.ordination import RDA, CCA
import pandas as pd

varechem = pd.read_table('varechem', sep=' ')
varespec = pd.read_table('varespec', sep=' ')

vare_rda = RDA(varespec.values, varechem.values,
               varespec.index, varespec.columns)
res = vare_rda.scores(1)
print res.eigvals
```
```
[ 137.34044141   95.8308358    48.56869977   33.09877583   24.84510216
   23.51820597   20.93992929   15.29185709   10.09267445    7.22888792
    5.94243135    4.61379848    4.05669       2.67820989   65.44012414
   45.10738814   29.63663915   20.57298507   17.18449793   15.57868306
   11.26704532   10.19726957    5.01205963]
```
Now, when I run the equivalent commands in R
```
library(vegan)
vare.rda <- rda(varespec, varechem)
vare.rda$CCA$eig
```
```
       RDA1        RDA2        RDA3        RDA4        RDA5        RDA6 
820.1042107 399.2847431 102.5616781  47.6316940  26.8382218  24.0480875 
       RDA7        RDA8        RDA9       RDA10       RDA11       RDA12 
 19.0643756  10.1669954   4.4287860   2.2720357   1.5353257   0.9255277 
      RDA13       RDA14 
  0.7155102   0.3118612
```
It looks like completely different output.  The same goes with the nearly all of the attributes calculated in RDA.

I agree, this is extremely weird since all of the tests passed.  Now the output for CCA is nearly completely consistent with vegan.  On a completely separate note, I'm not sure what to do about `site_constraints` within CCA, since there is no equivalent matrix in vegan's CCA output ...

jairideout commented:
Thanks for working on this @mortonjt @Jorge-C! Let me know if there's anything I can do to help.

mortonjt commented:
Just met up with @Jorge-C. Looks like the most of the RDA functionality is correct.  However, the calculation of the eigenvalues are wrong and @Jorge-C will submit a separate PR for that.

In the meantime, I'm going to aim to tidy up and commit all of my changes tonight.  I'll be out of town for 4 weeks starting Friday, so it'll be difficult for me to actively work on this.

mortonjt commented:
I just made the following changes

- Cleaned up the test cases for CA, CCA and RDA
- Removed the vegan test cases.  They are redundant and don't add anything new to the testing suite. Also, I'm not even sure if they would be BSD compatible
- Swapped out `np.linalg` with `scipy.linalg`.  They give identical results.  `scipy.linalg` will typically be faster for new users.

I'm still getting a weird error with sphinx - not sure what is going on there.
```
/Users/mortonjt/Documents/software/python/scikit-bio/doc/source/generated/generated/skbio.stats.distance.DissimilarityMatrix.rst:59: WARNING: toctree references unknown document u'generated/generated/skbio.stats.distance.DissimilarityMatrix.data'
```

Finally, @Jorge-C is investigating the bug with the miscalculated eigenvalues.  He'll be submitting to this PR in a bit.

ebolyen commented:
Did you `make clean` that's usually the cause when I see that error.
Although that path hasn't actually changed so it doesn't make a lot of
sense.

On Thu, Jun 18, 2015 at 5:00 PM, mortonjt <notifications@github.com> wrote:

> I just made the following changes
>
>    - Cleaned up the test cases for CA, CCA and RDA
>    - Removed the vegan test cases. They are redundant and don't add
>    anything new to the testing suite. Also, I'm not even sure if they would be
>    BSD compatible
>    - Swapped out np.linalg with scipy.linalg. They give identical
>    results. scipy.linalg will typically be faster for new users.
>
> I'm still getting a weird error with sphinx - not sure what is going on
> there.
>
> /Users/mortonjt/Documents/software/python/scikit-bio/doc/source/generated/generated/skbio.stats.distance.DissimilarityMatrix.rst:59: WARNING: toctree references unknown document u'generated/generated/skbio.stats.distance.DissimilarityMatrix.data'
>
> Finally, @Jorge-C <https://github.com/Jorge-C> is investigating the bug
> with the miscalculated eigenvalues. He'll be submitting to this PR in a bit.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/biocore/scikit-bio/pull/925#issuecomment-113322123>.
>

mortonjt commented:
Well shucks.  I forgot to run `make html`.  That fixed it.  Thanks @ebolyen !

josenavas commented:
@mortonjt look like there are merge conflicts

jairideout commented:
Thanks @mortonjt! Just so I know where to pick up from, what methods have been verified against the vegan test data besides RDA?

mortonjt commented:
I verified both CCA and RDA against vegan.  Checking CA right now ...

mortonjt commented:
I think CA is OK

When I use the same table in `skbio/stats/ordination/tests/data/L&L_CA_data` and run the following code in R
```
> X <- read.table('L&L_CA_data')
> vare.ca <- cca(X)
> scores(vare.ca, scaling=1)
```
It is equivalent to the test case in CA
```
        eigvals = pd.Series(np.array([0.09613302, 0.04094181]), self.pc_ids)
        # p. 458
        features = pd.DataFrame(np.array([[1.31871, -0.34374],  # V
                                          [-0.37215, 1.48150],
                                          [-0.99972, -0.92612]]),
                                self.feature_ids,
                                self.pc_ids)
        samples = pd.DataFrame(np.array([[-0.26322, -0.17862],  # F
                                         [-0.06835, 0.27211],
                                         [0.51685, -0.09517]]),
                               self.sample_ids,
                               self.pc_ids)
```
Same thing with `scaling=2`
```
> scores(vare.ca, scaling=2)
$species
          CA1         CA2
V1  0.4088694  0.06955181
V2 -0.1153860 -0.29976768
V3 -0.3099671  0.18739192

$sites
            CA1        CA2
sit1 -0.8489561  0.8827648
sit2 -0.2204587 -1.3448200
sit3  1.6669718  0.4703244
```

Here's the python test code for CA
```
        eigvals = pd.Series(np.array([0.09613302, 0.04094181]), self.pc_ids)
        # p. 460 L&L 1998
        features = pd.DataFrame(np.array([[0.40887, -0.06955],  # F_hat
                                          [-0.11539, 0.29977],
                                          [-0.30997, -0.18739]]),
                                self.feature_ids,
                                self.pc_ids)
        samples = pd.DataFrame(np.array([[-0.84896, -0.88276],  # V_hat
                                         [-0.22046, 1.34482],
                                         [1.66697, -0.47032]]),
                               self.sample_ids,
                               self.pc_ids)

```
And all of the test cases are passing

mortonjt commented:
But I'm not sure what's going on in the current travis run - I definitely didn't touch permanova ...

mortonjt commented:
Before I forget tomention,  my tests were done using vegan 2.0-8
On Jun 18, 2015 6:37 PM, "Jai Ram Rideout" <notifications@github.com> wrote:

> Thanks @mortonjt <https://github.com/mortonjt>! Just so I know where to
> pick up from, what methods have been verified against the vegan test data
> besides RDA?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/biocore/scikit-bio/pull/925#issuecomment-113329375>.
>

jairideout commented:
Thanks @mortonjt, this is really helpful!

Regarding the travis run, the tests were broken on master so updating this branch against latest master and fixing the merge conflicts should resolve the issue. I realize you're out of town and not actively working on this, just noting what needs to be done. Thanks again!

jairideout commented:
I created #1002 to track the RDA eigenvalue bug. Since this bug existed before the changes in this PR it isn't blocking a merge.

ElDeveloper commented:
Ready for review.

mortonjt commented on skbio/stats/ordination/_canonical_correspondence_analysis.py:
```diff
@@ -9,22 +9,25 @@
     the environment (or in other factors). It works from data on
-    abundances or counts of individuals and environmental variables,
-    and outputs ordination axes that maximize niche separation among
+    abundances or counts of samples and constrains variables,
```
should be "constrained variables"

mortonjt commented on skbio/stats/ordination/_canonical_correspondence_analysis.py:
```diff
@@ -9,22 +9,25 @@
-    abundances or counts of individuals and environmental variables,
-    and outputs ordination axes that maximize niche separation among
+    abundances or counts of samples and constrains variables,
+    and outputs ordination axes that maximize samples separation among
```
Glad that we dropped "niche".  May want to change this to "sample separation though"

mortonjt commented on skbio/stats/ordination/_correspondence_analysis.py:
```diff
@@ -33,158 +36,156 @@ class CA(Ordination):
     Parameters
     ----------
-    X : array_like
+    X : pd.DataFrame
```
Definitely want to change this to
```
X: pd.DataFrame
   Samples by features table (n, m)
```

mortonjt commented on skbio/stats/ordination/tests/test_canonical_correspondence_analysis.py:
```diff
@@ -0,0 +1,174 @@
+            columns=self.env_ids,
+            index=self.sample_ids
+            ).ix[:, :-1]
+
```
This was something that I noticed when modifying the CCA/RDA test cases.  @josenavas, do you think we can scrap the last column in that text file?  It doesn't seem like we need it.

josenavas replied:
Not sure. @Jorge-C Will be the best person to ask as it was like that before. I think it is used in some other test but I'm not 100% sure

Jorge-C replied:
I believe it can be dropped.

ElDeveloper replied:
Just removed it, let's see if the tests pass :O

On (Jul-13-15|16:26), Jorge Cañardo Alastuey wrote:
> > +                           'Sample4', 'Sample5', 'Sample6', 'Sample7',
> > +                           'Sample8', 'Sample9']
> > +        self.env_ids = ['Constraint0', 'Constraint1',
> > +                        'Constraint2', 'Constraint3']
> > +        self.pc_ids = ['CCA1', 'CCA2', 'CCA3', 'CCA4', 'CCA5', 'CCA6', 'CCA7',
> > +                       'CCA8', 'CCA9']
> > +        self.Y = pd.DataFrame(
> > +            np.loadtxt(get_data_path('example3_Y')),
> > +            columns=self.feature_ids,
> > +            index=self.sample_ids)
> > +        self.X = pd.DataFrame(
> > +            np.loadtxt(get_data_path('example3_X')),
> > +            columns=self.env_ids,
> > +            index=self.sample_ids
> > +            ).ix[:, :-1]
> > +
> 
> I believe it can be dropped.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/biocore/scikit-bio/pull/925/files#r34523618

ElDeveloper commented:
@mortonjt, just addressed your comments, let me know what else would you like me to change.

mortonjt commented on skbio/stats/ordination/_canonical_correspondence_analysis.py:
```diff
@@ -9,23 +9,25 @@
+    abundances or counts of samples and constraints variables,
+    and outputs ordination axes that maximize sample separation among species.
 
     It is better suited to extract the niches of taxa than linear
```
Just stating something that I should be addressed later, but not in this PR.

The term "niche" is incredibly ambiguous.  What exactly is the meaning here?
I'm not sure if there is an accepted standard definition for this term, and we may either want to clarify this or drop it in a future PR.

mortonjt commented:
Looks like the last check failed.  I really don't think it is worthwhile for this PR.  This is just too big ...

Noticed that `skbio/stats/ordination/tests/data/.DS_Store` is being added to this PR.  
Probably want to remove that binary.

Also don't forget about the CHANGELOG.md !

ElDeveloper commented:
Got it done, I just removed that commit and added a few notes to the
CHANGELOG.md (I am almost sure that's not the correct way to remove
existing functionality though ...).

On (Jul-13-15|17:22), mortonjt wrote:
> Looks like the last check failed.  I really don't think it is worthwhile for this PR.  This is just too big ...
> 
> Noticed that `skbio/stats/ordination/tests/data/.DS_Store` is being added to this PR.  
> Probably want to remove that binary.
> 
> Also don't forget about the CHANGELOG.md !
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/biocore/scikit-bio/pull/925#issuecomment-121097435

mortonjt commented:
Don't forget to resolve merge conflicts.

ElDeveloper commented:
Also, conflicts should be resolved :smile:

mortonjt commented:
I can't find any obvious problems.  @Jorge-C , @jairideout would you guys like to another look at this?

ebolyen commented:
I'm taking a full pass through this now.

ebolyen commented on skbio/stats/ordination/_canonical_correspondence_analysis.py:
```diff
@@ -9,23 +9,25 @@
 from skbio.util._decorator import experimental
 
 
-class CCA(Ordination):
```
I hate to say this, but under our new API lifecycle, the methods on these classes need to become `deprecated`. I know this commit was probably way in the past and might not be revertible, so if you have to copy it back in from master that is ok.

ebolyen replied:
This holds for all other removed Ordination classes.

ebolyen commented on skbio/_base.py:
```diff
@@ -23,3 +34,381 @@ class SkbioObject(with_metaclass(ABCMeta, object)):
+    biplot_scores : pd.DataFrame
+        Correlation coefficients of the samples with respect to the features.
+    sample_constraints : pd.DataFrame
+        A basis for the constrained ordination space. <- JORGE REVISE THIS
```
Commenting to highlight inline note.

Jorge-C replied:
I'd go with a definition from a book. For example, in "Numerical Ecology with R", by Daniel Borcard,Francois Gillet,Pierre Legendre:
> Site constraints (linear combinations of constraining variables): coordinates of the sites in the space of the explanatory variables X. These are the fitted site scores.

ebolyen replied:
Thanks @Jorge-C! @ElDeveloper please update.

gregcaporaso commented on CHANGELOG.md:
```diff
@@ -5,6 +5,18 @@
 ### Features
 * Added `to_regex` method to `skbio.sequence._iupac_sequence` ABC - it returns a regex object that matches all non-degenerate versions of the sequence.
 
+### Backward-incompatible changes
```
Change title to *Backward incompatible changes (experimental functionality)*

ElDeveloper commented:
Should be ready for another pass, thanks for the help @ebolyen, @gregcaporaso and @jairideout

ebolyen commented on skbio/io/format/ordination.py:
```diff
@@ -232,32 +233,32 @@ def _ordination_to_ordination_results(fh):
-        eigvals=eigvals, species=species, site=site, biplot=biplot,
-        site_constraints=cons, proportion_explained=prop_expl,
-        species_ids=species_ids, site_ids=site_ids)
+        short_method_name='', long_method_name='', eigvals=eigvals,
```
Is it worthwhile to allow users to set the `short_method_name` and `long_method_name` as kwargs in the reader?

mortonjt replied:
For the time being, I'd vote no - just because the file format is still a working progress.  
But I do think that this should be added in once we start hammering down the specifics of the file format - so we'll probably want to write up an issue for this.

jairideout replied:
Probably not, ideally we'll be phasing this file format out in favor of a non-lossy format.

ebolyen commented on skbio/stats/ordination/tests/test_correspondence_analysis.py:
```diff
@@ -0,0 +1,194 @@
+                                        ignore_directionality=True)
+
+    def test_maintain_chi_square_distance_scaling1(self):
+        """In scaling 1, chi^2 distance among rows (sites) is equal to
```
remove docstring. Make this a comment and change `(sites)` -> `(samples)`

ebolyen commented on skbio/stats/ordination/tests/test_correspondence_analysis.py:
```diff
@@ -0,0 +1,194 @@
+        npt.assert_almost_equal(chi2_distances, euclidean_distances)
+
+    def test_maintain_chi_square_distance_scaling2(self):
+        """In scaling 2, chi^2 distance among columns (species) is
```
make comment. `(species)` -> `(features)`

ebolyen commented on skbio/stats/ordination/tests/test_redundancy_analysis.py:
```diff
@@ -0,0 +1,171 @@
+            np.loadtxt(get_data_path(
+                'example2_biplot_scaling1')))
+
+        # These are wrong. @JorgeC don't forget to fix these
```
Unresolved TODO again

mortonjt replied:
That is a bug that has been there since the beginning and has been raised in #1002

I think we can take care of this problem after merging this PR

ebolyen replied:
Ok thanks, can we change the comment to reference that PR?

ebolyen commented on skbio/stats/ordination/tests/test_redundancy_analysis.py:
```diff
@@ -0,0 +1,171 @@
+            np.loadtxt(get_data_path(
+                'example2_biplot_scaling2')))
+
+        # These are wrong. @JorgeC don't forget to fix these
```
reference the issue same as above

ebolyen commented on skbio/_base.py:
```diff
@@ -23,3 +34,381 @@ class SkbioObject(with_metaclass(ABCMeta, object)):
+    """
+    default_write_format = 'ordination'
+
+    def __init__(self, short_method_name, long_method_name, eigvals,
```
tag each of these methods as experimental

ebolyen replied:
*public facing

mortonjt replied:
On it.  I'm assuming that we'll want to tag everything as experimental (i.e. pcoa, ca, cca, rda)

ebolyen replied:
yup! those other ones are actually already tagged, so it's just this class's methods which need to be tagged to my knowledge

ebolyen commented:
@ElDeveloper finished with review. I don't know the math, but everything else seemed nice.

ElDeveloper commented:
@ebolyen, I believe we (@mortonjt and me) have addressed all your comments.

mortonjt commented:
:+1:

ebolyen commented:
@ElDeveloper @mortonjt Looks great!

ElDeveloper commented:
:beers:

On (Jul-31-15| 8:51), Evan Bolyen wrote:
> @ElDeveloper @mortonjt Looks great!
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/biocore/scikit-bio/pull/925#issuecomment-126731262

mortonjt commented:
Nice!! Good work guys!

On Fri, Jul 31, 2015 at 10:51 AM, Yoshiki Vázquez Baeza <
notifications@github.com> wrote:

> :beers:
>
> On (Jul-31-15| 8:51), Evan Bolyen wrote:
> > @ElDeveloper @mortonjt Looks great!
> >
> > ---
> > Reply to this email directly or view it on GitHub:
> > https://github.com/biocore/scikit-bio/pull/925#issuecomment-126731262
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/biocore/scikit-bio/pull/925#issuecomment-126750341>.
>

gregcaporaso commented:
:beers: Thanks for the work on this one everyone!
